### PR TITLE
Add reviewing guidelines

### PR DESCRIPTION
**Pull request by lindsayad**
@idaholab/moose-ccb let's iterate on this. Please all feel free to push additions/changes to this PR. I'll keep this in draft for now as we're iterating, and eventually obviously we'll need a link to this page

### CODE REVIEW

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:7]:
```suggestion
```
this was informal discussion

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:21]: the gh api dont work here. There s no way to distinguish if you re doing both

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:25]:
```suggestion
- Reviews should be prompt. We should strive to provide a first quick review within XX business days
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:4]: let s categorize this more
& have some more basic guidelines

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:26]:
```suggestion
- New features should come under more scrutiny than re-factoring, which benefits from the confidence offered by the existing test suite

## Addressing Deadlock Between a CCB Author and CCB Reviewer
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:27]: maybe this should be a subsection and not its own big category. This doesnt happen often, and would happen even less with a tad more politeness

**Comment by joshuahansel** [comment]
This file captures some of the thoughts expressed in our discussion, but I think the review page should just be a list of specific instructions, rather than giving justification for the instructions. If we wanted to still have some of the discussion captured, we could maybe have a paragraph discussing the development philosophy.

**Comment by joshuahansel** [comment]
I'm not sure on the best way to iterate on this because I want to rewrite the whole page 😅 Here's my initial draft:
```
# Reviewing Pull Requests

MOOSE is a large project with large numbers of users and contributors. 
While MOOSE seeks the highest standards of software quality, it
is important to recognize that requiring perfection in proposed
code changes is not realistic, given the large amount of activity,
versus the limited availability of reviewers and time.
The guidelines outlined on this page are designed to guide
reviewers in finding the right balance between efficiency
and software quality.

## Conduct

- Be respectful.
- Try to be helpful. The reviewer is not expected to implement changes
  for the author, but the reviewer should try to ensure that
  the author has some direction, particularly if the author is a new 
  or infrequent contributor.
- Make it clear what are changes that you require, versus suggestions.
  For example, use "I suggest" for suggestions and imperative language,
  e.g., "Do this", for required changes.
- If it would be helpful, provide the author with a link to the contributing
  guidelines page.
- It is acceptable to make commits to the contributor's feature branch.
  Note that NQA-1 explicitly allows this, and an additional, independent
  review is only required if the additional commit(s) are significant.
- If the author refuses the requested changes, request an additional
  reviewer (ideally a `CODEOWNER` for the affected lines) to make a
  final decision on the changes.

## Review Strategy

- Review the high-level design of the changes before reviewing finer
  details.
- Consider the scope of the changes when determining how thorough
  the review should be: changes that potentially affect a large
  number of users should be subjected to greater scrutiny. Also
  consider the likelihood of additional capability being built
  on top of these changes.
- Carefully consider the user interface, such as input parameters,
  particularly for potentially high profile changes.
- Prioritize discovery of bugs that would manifest in very subtle,
  difficult-to-discover ways, rather than those bugs that
  would immediately become apparent, such as by an immediate error message.

## Testing

- The reviewer is not expected to test out the proposed
  changes, but this may be recommended in some cases when the 
  regression testing is deemed insufficient.
- At the very least, testing should cover most of the new code lines.
  Note that code coverage is automatically checked to meet a minimum
  value, but the reviewer may decide to override this failure if
  deemed justified. For example, a change may involve some branches
  in conditionals that should be impossible to hit.
- The reviewer may request additional testing.

## Documentation

- Inline code documentation such as doxygen is not required but may be requested
  when code is unclear.
- Request documentation for new classes and systems and also for 
  modifications to existing documentation where it is deemed important.
- Thorough documentation is not required, but it should be accurate
  and error-free, and it should clarify anything unusual or unclear.
- If the changes are significant enough to communicate to the 
  community, ensure there is a corresponding newsletter entry.

## Grammar, Spelling, and Typos

- Inline typos and grammar and spelling mistakes are recommended
  to be fixed, but are not as high of a priority as correctness in
  external documentation.
- Use Github suggestions for grammar, spelling, and typos.
```

**Comment by lindsayad** [comment]
I would say feel free to push this document @joshuahansel. I will have some edits but I like a lot of what you have

**Comment by loganharbour** [comment]
> I would say feel free to push this document @joshuahansel. I will have some edits but I like a lot of what you have

Just read through - I agree.

**Comment by cticenhour** [comment]
I agree as well. Thanks @joshuahansel!

**Comment by GiudGiud** [comment]
This is pretty much codifying a lower standard of review than I've been applying. I don't think I ll change my ways, and this should be checked for compliance with our SQA documents before we get it in. 

A few comments:
Checking out most PRs and checking that edge cases actually works is very helpful. You become the first user and find tons of bugs.
Fixing the spelling should be done as you're looking at the code, it's not that much and work and it will never get done later. 
Same for doxygen. Maybe add a note about "think about the next person trying to modify this code: is it commented enough?"
Or simply a blanket: always ask for more comments
The line coverage is easy to deceive, it s better to look at which features are being covered. 
There's also a few tentative contributors who are disrespectful / condescending, so the blanket "Be respectful" could be softened to allow some leniency. Alternatively we could have a system to close PRs for too much attitude from the tentative contributor
"- Prioritize discovery of bugs that would manifest in very subtle,
  difficult-to-discover ways, rather than those bugs that
  would immediately become apparent, such as by an immediate error message."
this is hard. What s easy is to think about to find bugs are the classic issues : higher order variables, distributed parallelism, nodal vs elemental quantities etc

Add a section about CIVET tips?
Add valgrind recipe if memory managements seems improvable
Add parallel sweeping recipes if convergence criteria are loose or the problem is inherently parallel

**Comment by GiudGiud** [comment]
We could add something about how reviews should be prompt ?
I dont like much that some PRs get ignored for weeks because everyone is busy. Maybe we should impose it on ourselves to get to every PR for a first opinion soon?

We should also add to this the rules for assignment of reviewers and for assignees (the merge button guy)

**Comment by lindsayad** [comment]
> This is pretty much codifying a lower standard of review than I've been applying. I don't think I ll change my ways, and this should be checked for compliance with our SQA documents before we get it in.

I'm hoping this is a document we can all eventually agree on. Let's work together to make it that way. I don't regard what @joshuahansel has written as the final version (or really anywhere close to the final version)

**Comment by GiudGiud** [comment]
We should also add something about how the reviewer has to understand what is going on and should feel free / be encouraged to ask questions to the PR maker.

**Comment by joshuahansel** [comment]
Ok, done. I added this issue you can reference, by the way: https://github.com/idaholab/moose/issues/23580.

**Comment by lindsayad** [comment]
I'm thinking I might close this

**Comment by joshuahansel** [comment]
I agree that we need to make some kind of decision here. I think it's useful to have some kind of review guidelines page, to give suggestions if nothing else, so I think it'd be a shame to throw it all away. I have no problem removing the content telling reviewers how thorough to be, if that's the uncomfortable bit with moving forward. Thoughts @GiudGiud?

**Comment by joshuahansel** [comment]
For example, what could definitely be kept are practical instructions about the rules and such (what situations require which reviewers, etc.). So it could give some of the information related to SQA requirements, and then additionally any practices that we'd like to impose on top of that.

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:10]:
```suggestion
Reviewers are bound by strict software quality standards, which they
enforce through careful reviews, often at the expense of their own research.
The guidelines outlined on this page are designed to help reviewers
provide an environment welcoming of external contributions and deal with
common situations. 
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:21]:
```suggestion
- Make it clear what are large changes that you require, versus suggestions.
  For example, use "I suggest" for suggestions and imperative language,
  e.g., "Do this", for required changes.
  Anything minor (typos, grammar, docstrings) is assumed to be required. If you
  do not wish for something to be fixed, do not suggest the change.
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:44]:
```suggestion
- Consider how the code will be understood outside of the context of a pull request. 
Are the comments and documentation enough to understand what is going on without
asking the initial developers.
```

**Comment by lindsayad** [comment]
Oh yea the rule that you don't need another reviewer if you push code that "isn't significant" is definitely worth having on the books

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:14]:
```suggestion
- Be respectful. Feel free to close the pull request if the contributor is not.
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:24]:
```suggestion
- Feel free to ask questions. It's normal to not understand everything in the pull request when you have
not written that code. Many times, the response to these questions should actually be added as comments
in the code.
- It is acceptable to make commits to the contributor's feature branch.
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:51]:
```suggestion
- At the very least, testing should cover all the features contributed, and most of the relevant code lines.
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:40]:
```suggestion
  If the proposed changes do not allow for additional capability, this can be a reason
  to refuse them.
- Carefully consider the user interface, such as input parameters,
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:61]:
```suggestion
  when code is unclear. When unclear whether it's unclear, ask for more documentation.
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:66]:
```suggestion
- When capabilities in the code change, it's likely user documentation will need to be adapted. Always check out
  the current state of the user documentation to make sure it does not become inaccurate.
- If the changes are significant enough to communicate to the
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:56]:
```suggestion
- The reviewer may request additional testing.
- The reviewer may enable additional testing recipes. Anything with manual memory management
  could benefit from a valgrind recipe. Anything with loose tolerances is likely to fail parallel testing,
  feel free to add those recipes. 
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:50]:
```suggestion
```
checking out the code locally and stress-testing it is straight up the best way to find issues. I dont think we should point the reviewer to not doing it

Comment (GiudGiud) [reply framework/doc/content/framework/reviewing.md:50]: "when the
  regression testing is deemed insufficient." this is true but very hard to detect!
It s easy to get coverage while completely missing the feature even, so catching edge cases is real hard

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:29]:
```suggestion
  final decision on the changes.
  - Be timely. Pull requests should only linger when there's good reasons for them not to move forward.
    If you cannot review, consider delegating the review to another member.
```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:30]:
```suggestion
- Be a team player. If there is a section of the code that you are not familiar with,
  feel free to ask for another reviewer's opinion.

```

**Review by GiudGiud — CHANGES_REQUESTED** [review]
If you don't want to loose your work I can understand that. Here are a few suggestions.

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [reply framework/doc/content/framework/reviewing.md:30]: maybe this isnt the most team player advice in the end. Maybe we should find something to say we should act as a team on this though. It's important.

**Comment by lindsayad** [comment]
> Let's debate this current set for a bit.

You mean the set of changes you just suggested?

**Comment by GiudGiud** [comment]
yup

File: framework/doc/content/framework/reviewing.md
Comment (joshuahansel) [inline framework/doc/content/framework/reviewing.md:57]:
```suggestion
  The comments and documentation should be enough to understand what is going on without
  asking the initial developers.
```

File: framework/doc/content/framework/reviewing.md
Comment (joshuahansel) [inline framework/doc/content/framework/reviewing.md:69]:
```suggestion
  could benefit from a valgrind recipe. Anything with loose tolerances is likely to fail parallel testing.
  Feel free to add those recipes. 
```

File: framework/doc/content/framework/reviewing.md
Comment (joshuahansel) [inline framework/doc/content/framework/reviewing.md:74]:
```suggestion
  when code is unclear. When in doubt, ask for more documentation.
```

File: framework/doc/content/framework/reviewing.md
Comment (joshuahansel) [inline framework/doc/content/framework/reviewing.md:80]:
```suggestion
- Ensure that existing documentation remains accurate with code changes.
```

**Review by joshuahansel — CHANGES_REQUESTED** [review]
I'm happy with @GiudGiud's changes - I just made some minor changes.

**Review by joshuahansel — DISMISSED** [review]

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [inline framework/doc/content/framework/reviewing.md:9]:
```suggestion

!alert note
We welcome external reviews of pull requests, even from parties with a stake in the proposed changes,
however due to the software quality requirements, pull requests will always require a review from a member
of the Change Control Board.

```

File: framework/doc/content/framework/reviewing.md
Comment (GiudGiud) [reply framework/doc/content/framework/reviewing.md:13]: the "reviewer may contribute to the PR" thing is also below

also contributon to PRs are easy for maintainers, but for non-maintainers it s a process right?
either gh suggestions or make a PR to the PR on the origin fork

Comment (joshuahansel) [reply framework/doc/content/framework/reviewing.md:13]: Ok, changed the suggestion. How's that?

**Review by lindsayad — COMMENTED** [review]
I approve this message

File: framework/doc/content/framework/reviewing.md
Comment (lindsayad) [inline framework/doc/content/framework/reviewing.md:24]:
```suggestion
  Anything minor (typos, grammar, docstrings) is assumed to be required. If you
  do not wish for something to be changed, do not suggest the change.
```

File: framework/doc/content/framework/reviewing.md
Comment (cticenhour) [inline framework/doc/content/framework/reviewing.md:15]:
```suggestion
- The MOOSE project strives to create a respectful environment for all contributors, reviewers, and users. Therefore, please be respectful during the pull request submission and approval process. Pull requests can be closed if interactions do not meet this standard. 
```

File: framework/doc/content/framework/reviewing.md
Comment (cticenhour) [inline framework/doc/content/framework/reviewing.md:20]:
```suggestion
- Make it clear the large changes that are required by the reviewer, versus suggestions.
```

Comment (GiudGiud) [reply framework/doc/content/framework/reviewing.md:20]: Clearly state required large changes ?

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
only squashed some commits

**Comment by lindsayad** [comment]
thanks team!

**Pull request merged**
